Cluster code no longer resolves hostnames into IPs.

rickardoberg opened the pull request:
More changes to allow cross-datacenter clusters.

rickardoberg committed:
Cluster code no longer resolves hostnames into IPs.

More changes to allow cross-datacenter clusters.

digitalstain commented on enterprise/cluster/src/main/java/org/neo4j/cluster/statemachine/StateTransitionLogger.java:
```diff
@@ -53,7 +53,7 @@ public void stateTransition( StateTransition transition )
         Log log = logProvider.getLog( transition.getOldState().getClass() );
 
-        if ( log.isDebugEnabled() )
+//        if ( log.isDebugEnabled() )
```
Why is this commented out? The rational was, as far as i understand it, to not create that big string message to log if it was going to be discarded anyway. Is the runtime cost assumption invalidated somehow? If that's the case, then we should remove the block altogether.

digitalstain commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java:
```diff
@@ -432,8 +447,19 @@ private void catchUpWithMaster() throws IllegalArgumentException, InterruptedExc
         // Unpause the update puller, because we know that we are a slave that just started communication with master.
-        updatePuller.unpause();
-        updatePuller.await( UpdatePuller.NEXT_TICKET, true );
+        while ( true )
```
This loop has no guarantee that it will terminate. We've had issues with similar constructs lately and i am a bit sceptical as to their safety. I'd prefer either logging failures with exponential back off on retries, or a bound number of iterations and raising a failure if not successful.

rickardoberg replied:
It is guaranteed to terminate. Unpause will always work within a very short while.

digitalstain commented on enterprise/ha/src/test/java/org/neo4j/test/ha/ClusterManager.java:
```diff
@@ -966,7 +966,7 @@ private void startMember( InstanceId serverId ) throws URISyntaxException, IOExc
                 builder.setConfig( ClusterSettings.initial_hosts, initialHosts.toString() );
                 builder.setConfig( ClusterSettings.server_id, serverId + "" );
-                builder.setConfig( ClusterSettings.cluster_server, "0.0.0.0:" + clusterPort );
+                builder.setConfig( ClusterSettings.cluster_server, clusterUri.getHost() + ":" + clusterPort );
```
I am wondering if this is necessary. The tests should continue to work with ips regardless of this change, _requiring_ hostnames is a behavioral change which is not expected.

lutovich commented on enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/cluster/ClusterState.java:
```diff
@@ -299,9 +299,11 @@
-                                        StringBuffer errorMessage = new StringBuffer( "Failed to join cluster because I saw two instances with the same ServerId" );
-                                        errorMessage.append( "One is " ).append( discoveredInstance.getJoiningId() );
-                                        errorMessage.append( " The other is " ).append( configurationRequested );
+                                        StringBuffer errorMessage = new StringBuffer(
```
StringBuffer is probably an overkill. Simple string concatenation should be equally performant and more readable.

lutovich commented on enterprise/ha/src/test/java/org/neo4j/test/ha/ClusterRule.java:
```diff
@@ -112,17 +114,15 @@ public ClusterRule availabilityChecks( List<Predicate<ManagedCluster>> checks )
+    @Override
+    protected void before() throws Throwable
+    {
+        ClusterRule.this.storeDirectory = testDirectory.directory( methodName );
```
Probably `ClusterRule.this.` part is not needed

rickardoberg committed:
Try to resend messages if channel was closed

Move all state switchers to the HAMS thread, to make it more deterministic. This should avoid rare race conditions where those things happen in the wrong order.

Allow state change where wrong master becomes available, as there might have been a new election during the master switch.

davidegrohmann commented:
subsumed by https://github.com/neo4j/neo4j/pull/5209 for the hostname part, the other fixes are gonna backported to 2.2 first and then forward merged to 2.3
